lib/ansible/modules/database/postgresql/postgresql_user.py

```diff
@@ -750,6 +763,91 @@ def get_valid_flags_by_version(cursor):
     ]
 
 
+class PgMembership():
+    def __init__(self, module, cursor, target_roles, groups, fail_on_role=True):
+        self.module = module
+        self.cursor = cursor
+        self.target_roles = [r.strip() for r in target_roles]
+        self.groups = groups
+        self.granted = {}
+        self.fail_on_role = fail_on_role
+        self.non_existent_roles = []
+        self.changed = False
+        self.__check_roles_exist()
+
+    def grant(self):
+        for group in self.groups:
+            self.granted[group] = []
+
+            for role in self.target_roles:
+                # If role is in a group now, pass:
+                if self.__check_membership(group, role):
```

Comment:
(Or maybe I misunderstood something, but this can probably be done more efficiently by doing at most one query per outer loop iteration.)

Reply:
the idea sounds good. However this class is almost the same as in postgresql_membership module. I'm going to move it to module_utils/postgresql.py and make it shared.
However I'd like to do it by particular PR.
Your idea is good, be sure i'll try to make this optimization soon for both modules at the same time (if i do it now it might lead me to confusion in the future and i'll have to rewrite it again)

Reply:
Sure, sounds good to me :)